src/main/java/sg/dex/starfish/impl/memory/AMemoryOperation.java

```diff
@@ -30,19 +30,22 @@ public Job invokeAsync(Map<String, Object> params) {
 		// pool)
 		// But this should be sufficient for most cases.
 		final CompletableFuture<Map<String, Object>> future = new CompletableFuture<>();
-
+		MemoryJob memoryJob = MemoryJob.create(future);
+        memoryJob.setStatus(Job.Status.scheduled);
 		MemoryAgent.THREAD_POOL.submit(() -> {
 			try {
 				Map<String, Object> result = compute(params);
-				future.complete(result); // success
+				if(future.complete(result)){
+                    memoryJob.setStatus(Job.Status.succeeded);
+                }
 			}
 			catch (Throwable t) {
-				future.completeExceptionally(t); // failure
+				future.completeExceptionally(t);
+				memoryJob.setStatus(Job.Status.failed);// failure
 			}
-			assert (future.isDone());
 		});
 
-		MemoryJob memoryJob = MemoryJob.create(future);
+		memoryJob.setStatus(Job.Status.running);
```

Comment:
Race condition? Job could already have completed? Must set the status to running immediately before executing the Job!